### PR TITLE
Initial epic implementation

### PR DESCRIPTION
**Pull request by sachiniSam**
## Automation tests
 - Integration tests
   > Integration tests are carried out in – src/test

Ubuntu 16.04
WSO2 EI 6.3.0

### CODE REVIEW

File: src/test/INTEGRATION_TEST.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+		<messageFormatter contentType="application/json+fhir" class="org.apache.synapse.commons.json.JsonStreamFormatter"/>
+		<messageBuilder contentType="application/json+fhir" class="org.apache.synapse.commons.json.JsonStreamBuilder"/>
+
+	3. Extract the certificate from browser by navigating to the official epic sandbox URL and Add the certificate obtained to "<EI_HOME>/repository/.
+
+	4. Update connector properties file "Epic.properties" located in "EPIC_HOME/src/test/resources/artifacts/ESB/connector/config/",
+	with relevant information
+
+    5.Navigate to "{EPIC_HOME}" and run the following command.
+		$ mvn clean install -Dskip-tests=false
```
Comment (Biruntha) [inline src/test/INTEGRATION_TEST.txt]: Can you add description for the properties used in Epic.properties file?

File: src/test/INTEGRATION_TEST.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+	Steps to follow in setting integration test.
+
+	1. Download WSO2 EI 6.3.0 from official website and copy the EI 6.3.0 zip to "{EPIC_HOME}/repository".
+
+	2. Enable the following message formatter and message builder in the WSO2 EI axis2.xml
+		<messageFormatter contentType="application/json+fhir" class="org.apache.synapse.commons.json.JsonStreamFormatter"/>
+		<messageBuilder contentType="application/json+fhir" class="org.apache.synapse.commons.json.JsonStreamBuilder"/>
+
+	3. Extract the certificate from browser by navigating to the official epic sandbox URL and Add the certificate obtained to "<EI_HOME>/repository/.
```
Comment (Biruntha) [inline src/test/INTEGRATION_TEST.txt]: Can you add sample URL here?

File: .gitignore [hunk 3]
```diff
@@ -21,3 +21,9 @@
 
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+
+.idea/
+org.wso2.carbon.esb.connector.epic.iml
+repository/wso2ei-6.3.0.zip
+repository/open-icepiccom.crt
+target/
```
Comment (Megala21) [inline .gitignore]: Please add an extra line

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+            <name>WSO2 Release Distribution Repository</name>
+            <url>http://maven.wso2.org/nexus/service/local/staging/deploy/maven2/</url>
+        </repository>
+        <snapshotRepository>
+            <id>wso2.snapshots</id>
+            <name>Apache Snapshot Repository</name>
+            <url>http://maven.wso2.org/nexus/content/repositories/snapshots/</url>
+        </snapshotRepository>
+    </distributionManagement>
+</project>
```
Comment (Megala21) [inline pom.xml]: Please add an extra line. Please correct in other places as well.

File: src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import org.wso2.connector.integration.test.base.ConnectorIntegrationTestBase;
+import org.wso2.connector.integration.test.base.RestResponse;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.io.IOException;
+
```
Comment (Megala21) [inline src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java]: Please remove the extra line

File: src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    /**
+     * Positive test case for getConformance method with mandatory parameters.
+     */
+    @Test(enabled = true, description = "epic {getConformance} integration test with mandatory parameters.")
+    public void getConformanceWithMandatoryParameters() throws IOException, JSONException {
+        eiRequestHeadersMap.put("Action", "urn:getConformance");
+        RestResponse<JSONObject> eiRestResponse = sendJsonRestRequest(proxyUrl, "POST", eiRequestHeadersMap, "getConformanceMandatory.json");
+        final String apiEndPoint = connectorProperties.getProperty("base") + "/metadata";
+        RestResponse<JSONObject> apiRestResponse = sendJsonRestRequest(apiEndPoint, "GET", apiRequestHeadersMap);
+        Assert.assertEquals(eiRestResponse.getHttpStatusCode(), 200);
```
Comment (Megala21) [inline src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java]: Can we add a description, on why this test case is failing.

File: src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    /**
+     * Positive test case for readSpecificResourceById method with Mandatory parameters.
+     */
+    @Test(enabled = true, description = "epic {readSpecificResourceById} integration test with Mandatory parameters.")
+    public void readSpecificResourceByIdWithMandatoryParameters() throws IOException, JSONException {
+        eiRequestHeadersMap.put("Action", "urn:readById");
+        RestResponse<JSONObject> eiRestResponse = sendJsonRestRequest(proxyUrl, "POST", eiRequestHeadersMap, "readSpecificResourceById.json");
+        final String apiEndPoint = connectorProperties.getProperty("base") + "/" + connectorProperties.getProperty("type") + "/" + connectorProperties.getProperty("id");
```
Comment (Megala21) [inline src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java]: Please follow 120 limit. Seems it is going beyond that,please fix in all places

File: src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    public void searchMedicationOrderwithOptionalParameters() throws IOException, JSONException {
+        eiRequestHeadersMap.put("Action", "urn:searchMedicationOrder");
+        RestResponse<JSONObject> eiRestResponse = sendJsonRestRequest(proxyUrl, "POST", eiRequestHeadersMap, "searchMedicationOrder.json");
+        final String apiEndPoint = connectorProperties.getProperty("base") + "/" + connectorProperties.getProperty("medicationOrderType") +
+                "?patient=" + connectorProperties.getProperty("patientId") + "&status=" + connectorProperties.getProperty("status");
+        RestResponse<JSONObject> apiRestResponse = sendJsonRestRequest(apiEndPoint, "GET", apiRequestHeadersMap);
+        Assert.assertEquals(eiRestResponse.getHttpStatusCode(), 200);
+        Assert.assertEquals(apiRestResponse.getHttpStatusCode(), 200);
+    }
+
```
Comment (Megala21) [inline src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java]: Please remove unwanted extra lines.

File: src/test/INTEGRATION_TEST.txt [hunk 9]
```diff
@@ -0,0 +1,56 @@
+﻿Product: Integration tests for WSO2 ESB epic connector through REST API
```
Comment (Biruntha) [inline src/test/INTEGRATION_TEST.txt]: can you update this file as INTEGRATION_TEST.md

File: docs/operation.md [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+**Sample request**
+
+Following is a sample REST request that can be handled by the readById operation.
+```
+{
+ "base": “https://open-ic.epic.com/FHIR/api/FHIR/DSTU2/",
+ "type": “Binary",
+ "id": "T51iWjMg3MhfQWNRaJpmEF8ALTC6lQcunX-VTNpxqyTiV8.6oSleCNUa5Yr82YE4CJaCO0rcZankXUhpoCWkzlwB"
+}
+```
```
Comment (Biruntha) [inline docs/operation.md]: Add sample response for each operations as in [1].
[1] https://docs.wso2.com/display/ESBCONNECTORS/Working+with+Sheet+Operations

File: src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+     */
+    @Test(enabled = true, description = "epic {readSpecificResourceById} integration test with Mandatory parameters.")
+    public void readSpecificResourceByIdWithMandatoryParameters() throws IOException, JSONException {
+        eiRequestHeadersMap.put("Action", "urn:readById");
+        RestResponse<JSONObject> eiRestResponse = sendJsonRestRequest(proxyUrl, "POST", eiRequestHeadersMap, "readSpecificResourceById.json");
+        final String apiEndPoint = connectorProperties.getProperty("base") + "/"
+                + connectorProperties.getProperty("type") + "/" + connectorProperties.getProperty("id");
+        RestResponse<JSONObject> apiRestResponse = sendJsonRestRequest(apiEndPoint, "GET", apiRequestHeadersMap);
+        Assert.assertEquals(eiRestResponse.getHttpStatusCode(), 200,"Invalid parameter value");
+        Assert.assertEquals(apiRestResponse.getHttpStatusCode(), 200, "Invalid parameter value");
```
Comment (Biruntha) [inline src/test/java/org/wso2/carbon/connector/integration/test/Epic/EpicConnectorIntegrationTest.java]: Can we assert more properties from response body ? fix other test cases as well.

refer [1].
[1] https://github.com/wso2-extensions/esb-connector-salesforcerest/blob/master/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java#L118

File: docs/README.md [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# Epic EI Connector
+The Epic [connector](https://docs.wso2.com/display/EI611/Working+with+Connectors) allows you to access the Epic FHIR APIs through the WSO2 EI. 
+Epic connector currently supports the DSTU 2 Final (1.0.2) version of the FHIR standard.
```
Comment (Biruntha) [inline docs/README.md]: Mention which authorization mechanism you are using and provide the api doc as well

File: src/test/resources/artifacts/ESB/connector/config/Epic.properties [hunk 13]
```diff
@@ -0,0 +1,33 @@
+proxyDirectoryRelativePath=/artifacts/ESB/config/proxies/Epic/
+requestDirectoryRelativePath=/artifacts/ESB/config/restRequests/Epic/
+
```
Comment (Biruntha) [inline src/test/resources/artifacts/ESB/connector/config/Epic.properties]: Currently we are automating the integration test case to minimize manual modifications. So could you pls do the needful to automate this.

File: src/test/resources/artifacts/ESB/connector/config/Epic.properties [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+base=https://open-ic.epic.com/FHIR/api/FHIR/DSTU2
```
Comment (Biruntha) [inline src/test/resources/artifacts/ESB/connector/config/Epic.properties]: Move all the properties that are specific to accounts to separate properties file inside "connector/repository/" folder .
Refer [1].

[1]https://github.com/wso2-extensions/esb-connector-googlespreadsheet/pull/13/files#diff-224e68342d61e85c9386bf00d3afe9dfR1

**Pull request merged**